File: apps/backend/src/config/gsm-config.ts
/**
 * My Hibachi Backend Configuration Loader
 * Loads secrets from GSM with environment variable fallbacks
 */

import { getGSMClient, createBackendAPIClient } from '../../../../libs/gsm-client/src/index';

export interface BackendConfig {
  // Database
  DB_URL: string;
  
  // Authentication
  JWT_SECRET: string;
  JWT_REFRESH_SECRET: string;
  SESSION_ENCRYPTION_KEY: string;
  
  // External APIs
  STRIPE_SECRET_KEY: string;
  STRIPE_WEBHOOK_SECRET: string;
  OPENAI_API_KEY: string;
  GOOGLE_MAPS_SERVER_KEY: string;
  
  // Communication
  SENDGRID_API_KEY: string;
  EMAIL_DEFAULT_FROM: string;
  
  // Internal
  REDIS_URL: string;
  ADMIN_PANEL_SECRET: string;
  
  // Feature Flags
  FF_ENABLE_AI_BOOKING_V3: boolean;
  FF_ENABLE_TRAVEL_FEE_V2: boolean;
  FF_ENABLE_SMS_REMINDERS: boolean;
  
  // Monitoring
  CONFIG_VERSION: string;
  NODE_ENV: string;
}

class ConfigManager {
  private config: BackendConfig | null = null;
  private environment: 'dev' | 'stg' | 'prod';
  private configVersion: string | null = null;
  
  constructor() {
    // Auto-detect environment
    this.environment = this.detectEnvironment();
    console.log(`🏗️ Initializing config for environment: ${this.environment}`);
  }

  /**
   * Load configuration from GSM with env fallbacks
   */
  async loadConfig(): Promise<BackendConfig> {
    console.log('📡 Loading configuration...');
    
    try {
      const gsmClient = createBackendAPIClient(this.environment);
      
      // Load all secrets
      const secrets = await gsmClient.getSecrets([
        // Database
        'DB_URL',
        
        // Auth
        'JWT_SECRET',
        'JWT_REFRESH_SECRET', 
        'SESSION_ENCRYPTION_KEY',
        
        // External APIs (from global)
        'STRIPE_SECRET_KEY',
        'STRIPE_WEBHOOK_SECRET',
        'OPENAI_API_KEY',
        'GOOGLE_MAPS_SERVER_KEY',
        'SENDGRID_API_KEY',
        
        // Communication
        'EMAIL_DEFAULT_FROM',
        
        // Internal
        'REDIS_URL',
        'ADMIN_PANEL_SECRET',
        
        // Feature Flags
        'FF_ENABLE_AI_BOOKING_V3',
        'FF_ENABLE_TRAVEL_FEE_V2', 
        'FF_ENABLE_SMS_REMINDERS',
        
        // Monitoring
        'CONFIG_VERSION'
      ]);

      this.config = {
        // Database
        DB_URL: secrets.DB_URL,
        
        // Authentication
        JWT_SECRET: secrets.JWT_SECRET,
        JWT_REFRESH_SECRET: secrets.JWT_REFRESH_SECRET,
        SESSION_ENCRYPTION_KEY: secrets.SESSION_ENCRYPTION_KEY || this.generateFallbackKey(),
        
        // External APIs
        STRIPE_SECRET_KEY: secrets.STRIPE_SECRET_KEY,
        STRIPE_WEBHOOK_SECRET: secrets.STRIPE_WEBHOOK_SECRET,
        OPENAI_API_KEY: secrets.OPENAI_API_KEY,
        GOOGLE_MAPS_SERVER_KEY: secrets.GOOGLE_MAPS_SERVER_KEY,
        
        // Communication
        SENDGRID_API_KEY: secrets.SENDGRID_API_KEY,
        EMAIL_DEFAULT_FROM: secrets.EMAIL_DEFAULT_FROM || 'noreply@myhibachichef.com',
        
        // Internal
        REDIS_URL: secrets.REDIS_URL || 'redis://localhost:6379',
        ADMIN_PANEL_SECRET: secrets.ADMIN_PANEL_SECRET,
        
        // Feature Flags (with safe defaults)
        FF_ENABLE_AI_BOOKING_V3: this.parseBoolean(secrets.FF_ENABLE_AI_BOOKING_V3, true),
        FF_ENABLE_TRAVEL_FEE_V2: this.parseBoolean(secrets.FF_ENABLE_TRAVEL_FEE_V2, false),
        FF_ENABLE_SMS_REMINDERS: this.parseBoolean(secrets.FF_ENABLE_SMS_REMINDERS, true),
        
        // Monitoring
        CONFIG_VERSION: secrets.CONFIG_VERSION || '1',
        NODE_ENV: process.env.NODE_ENV || this.environment
      };

      this.configVersion = this.config.CONFIG_VERSION;
      
      console.log('✅ Configuration loaded successfully');
      console.log(`📊 Config version: ${this.configVersion}`);
      console.log(`🚀 Feature flags: AI_V3=${this.config.FF_ENABLE_AI_BOOKING_V3}, Travel_V2=${this.config.FF_ENABLE_TRAVEL_FEE_V2}`);
      
      return this.config;
    } catch (error) {
      console.error('❌ Failed to load config from GSM:', error);
      console.log('🔄 Falling back to environment variables...');
      
      // Fallback to environment variables
      return this.loadFromEnv();
    }
  }

  /**
   * Get current configuration
   */
  getConfig(): BackendConfig {
    if (!this.config) {
      throw new Error('Configuration not loaded. Call loadConfig() first.');
    }
    return this.config;
  }

  /**
   * Check if config version changed and reload if needed
   */
  async checkAndReload(): Promise<boolean> {
    if (!this.config) return false;
    
    try {
      const gsmClient = createBackendAPIClient(this.environment);
      const hasChanged = await gsmClient.checkConfigVersion();
      
      if (hasChanged) {
        console.log('🔄 Config version changed, reloading...');
        await this.loadConfig();
        return true;
      }
      
      return false;
    } catch (error) {
      console.warn('⚠️ Could not check config version:', error);
      return false;
    }
  }

  /**
   * Get feature flag value
   */
  getFeatureFlag(flag: keyof BackendConfig): boolean {
    const config = this.getConfig();
    return Boolean(config[flag]);
  }

  /**
   * Start background config monitoring
   */
  startConfigMonitoring(intervalMinutes: number = 5): void {
    console.log(`🔍 Starting config monitoring (every ${intervalMinutes} minutes)`);
    
    setInterval(async () => {
      try {
        const changed = await this.checkAndReload();
        if (changed) {
          console.log('🔄 Configuration reloaded due to version change');
        }
      } catch (error) {
        console.error('❌ Config monitoring error:', error);
      }
    }, intervalMinutes * 60 * 1000);
  }

  // Private methods
  private detectEnvironment(): 'dev' | 'stg' | 'prod' {
    const env = process.env.NODE_ENV?.toLowerCase();
    const branch = process.env.VERCEL_GIT_COMMIT_REF || process.env.BRANCH_NAME;
    
    if (env === 'production' || branch === 'main') return 'prod';
    if (env === 'staging' || branch === 'dev') return 'stg';
    return 'dev';
  }

  private parseBoolean(value: string | undefined, defaultValue: boolean): boolean {
    if (!value) return defaultValue;
    return value.toLowerCase() === 'true' || value === '1';
  }

  private generateFallbackKey(): string {
    return process.env.SESSION_ENCRYPTION_KEY || 'fallback-key-' + Math.random().toString(36);
  }

  private loadFromEnv(): BackendConfig {
    console.log('📁 Loading configuration from environment variables');
    
    return {
      // Database
      DB_URL: process.env.DB_URL || 'postgresql://localhost/myhibachi_dev',
      
      // Authentication
      JWT_SECRET: process.env.JWT_SECRET || 'dev-jwt-secret',
      JWT_REFRESH_SECRET: process.env.JWT_REFRESH_SECRET || 'dev-refresh-secret',
      SESSION_ENCRYPTION_KEY: process.env.SESSION_ENCRYPTION_KEY || 'dev-session-key',
      
      // External APIs
      STRIPE_SECRET_KEY: process.env.STRIPE_SECRET_KEY || '',
      STRIPE_WEBHOOK_SECRET: process.env.STRIPE_WEBHOOK_SECRET || '',
      OPENAI_API_KEY: process.env.OPENAI_API_KEY || '',
      GOOGLE_MAPS_SERVER_KEY: process.env.GOOGLE_MAPS_SERVER_KEY || '',
      
      // Communication
      SENDGRID_API_KEY: process.env.SENDGRID_API_KEY || '',
      EMAIL_DEFAULT_FROM: process.env.EMAIL_DEFAULT_FROM || 'noreply@myhibachichef.com',
      
      // Internal
      REDIS_URL: process.env.REDIS_URL || 'redis://localhost:6379',
      ADMIN_PANEL_SECRET: process.env.ADMIN_PANEL_SECRET || 'dev-admin-secret',
      
      // Feature Flags
      FF_ENABLE_AI_BOOKING_V3: this.parseBoolean(process.env.FF_ENABLE_AI_BOOKING_V3, true),
      FF_ENABLE_TRAVEL_FEE_V2: this.parseBoolean(process.env.FF_ENABLE_TRAVEL_FEE_V2, false),
      FF_ENABLE_SMS_REMINDERS: this.parseBoolean(process.env.FF_ENABLE_SMS_REMINDERS, true),
      
      // Monitoring
      CONFIG_VERSION: process.env.CONFIG_VERSION || '1',
      NODE_ENV: process.env.NODE_ENV || 'development'
    };
  }
}

// Singleton instance
const configManager = new ConfigManager();

/**
 * Load and get configuration
 */
export async function loadConfig(): Promise<BackendConfig> {
  return await configManager.loadConfig();
}

/**
 * Get current configuration (must be loaded first)
 */
export function getConfig(): BackendConfig {
  return configManager.getConfig();
}

/**
 * Check for config updates and reload if needed
 */
export async function reloadConfigIfChanged(): Promise<boolean> {
  return await configManager.checkAndReload();
}

/**
 * Start background config monitoring
 */
export function startConfigMonitoring(intervalMinutes?: number): void {
  configManager.startConfigMonitoring(intervalMinutes);
}

/**
 * Get feature flag value
 */
export function getFeatureFlag(flag: keyof BackendConfig): boolean {
  return configManager.getFeatureFlag(flag);
}

/**
 * Convenience functions for common feature flags
 */
export const isAIBookingV3Enabled = () => getFeatureFlag('FF_ENABLE_AI_BOOKING_V3');
export const isTravelFeeV2Enabled = () => getFeatureFlag('FF_ENABLE_TRAVEL_FEE_V2');
export const isSMSRemindersEnabled = () => getFeatureFlag('FF_ENABLE_SMS_REMINDERS');

File: apps/backend/src/config/gsm_config.py
"""
Google Secret Manager Configuration Module for My Hibachi Backend
Enterprise-grade secret loading with fallbacks and caching
"""

import os
import logging
from typing import Dict, Any, Optional
from functools import lru_cache
import asyncio
from datetime import datetime, timedelta, timezone

try:
    from google.cloud import secretmanager

    GSM_AVAILABLE = True
except ImportError:
    GSM_AVAILABLE = False
    secretmanager = None

logger = logging.getLogger(__name__)


class GSMConfig:
    """Google Secret Manager configuration loader with fallbacks"""

    def __init__(self, project_id: str = "my-hibachi-crm", environment: str = "prod"):
        self.project_id = project_id
        self.environment = environment
        self.client = None
        self._cache: Dict[str, Any] = {}
        self._cache_expiry: Dict[str, datetime] = {}
        self._cache_ttl = timedelta(minutes=15)  # Cache for 15 minutes

        if GSM_AVAILABLE:
            try:
                self.client = secretmanager.SecretManagerServiceClient()
                logger.info(f"✅ GSM client initialized for project: {project_id}")
            except Exception as e:
                logger.warning(f"⚠️ GSM client initialization failed: {e}")
                self.client = None
        else:
            logger.warning(
                "⚠️ Google Cloud Secret Manager not available, using environment variables only"
            )

    def _build_secret_name(self, category: str, key: str) -> str:
        """Build the full secret name for GSM"""
        return f"{self.environment}-{category}-{key}"

    def _is_cache_valid(self, secret_name: str) -> bool:
        """Check if cached value is still valid"""
        if secret_name not in self._cache_expiry:
            return False
        return datetime.now(timezone.utc) < self._cache_expiry[secret_name]

    async def get_secret(
        self, category: str, key: str, fallback_env_var: Optional[str] = None
    ) -> Optional[str]:
        """
        Get secret value with GSM + environment variable fallback

        Args:
            category: Secret category (global, backend-api, frontend-web, frontend-admin)
            key: Secret key name
            fallback_env_var: Environment variable name for fallback

        Returns:
            Secret value or None if not found
        """
        secret_name = self._build_secret_name(category, key)

        # Check cache first
        if self._is_cache_valid(secret_name):
            return self._cache[secret_name]

        # Try GSM first
        if self.client:
            try:
                secret_path = f"projects/{self.project_id}/secrets/{secret_name}/versions/latest"
                response = await asyncio.get_event_loop().run_in_executor(
                    None, self.client.access_secret_version, {"name": secret_path}
                )
                # Strip whitespace to handle Windows line endings (\r\n) in secret values
                value = response.payload.data.decode("UTF-8").strip()

                # Cache the result
                self._cache[secret_name] = value
                self._cache_expiry[secret_name] = datetime.now(timezone.utc) + self._cache_ttl

                logger.debug(f"✅ Retrieved secret from GSM: {secret_name}")
                return value

            except Exception as e:
                logger.warning(f"⚠️ Failed to retrieve {secret_name} from GSM: {e}")

        # Fallback to environment variable
        env_var = fallback_env_var or key
        env_value = os.getenv(env_var)
        if env_value:
            # Cache the fallback value
            self._cache[secret_name] = env_value
            self._cache_expiry[secret_name] = datetime.now(timezone.utc) + self._cache_ttl
            logger.debug(f"✅ Retrieved secret from environment: {env_var}")
            return env_value

        logger.error(f"❌ Secret not found: {secret_name} (env: {env_var})")
        return None

    async def get_secrets_batch(self, secrets: list) -> Dict[str, str]:
        """
        Get multiple secrets in batch

        Args:
            secrets: List of tuples (category, key, optional_env_var)

        Returns:
            Dictionary of {key: value}
        """
        results = {}
        tasks = []

        for secret_info in secrets:
            if len(secret_info) == 2:
                category, key = secret_info
                env_var = None
            else:
                category, key, env_var = secret_info

            task = self.get_secret(category, key, env_var)
            tasks.append((key, task))

        # Execute all requests concurrently
        for key, task in tasks:
            try:
                value = await task
                if value:
                    results[key] = value
            except Exception as e:
                logger.error(f"Failed to retrieve secret {key}: {e}")

        return results

    def clear_cache(self):
        """Clear the secret cache"""
        self._cache.clear()
        self._cache_expiry.clear()
        logger.info("🗑️ Secret cache cleared")


# Global instance
_gsm_config: Optional[GSMConfig] = None


@lru_cache()
def get_gsm_config() -> GSMConfig:
    """Get singleton GSM configuration instance.

    Environment Variables:
        GCP_PROJECT_ID: Google Cloud project ID (default: my-hibachi-crm)
        GSM_ENVIRONMENT: GSM secret prefix (default: prod)
            - Use 'prod' for production secrets (prod-backend-api-*)
            - Use 'staging' for staging secrets (staging-backend-api-*)
            - This is SEPARATE from the ENVIRONMENT variable used by Settings/Pydantic

    Note:
        GSM_ENVIRONMENT controls the secret naming prefix in GSM.
        ENVIRONMENT controls the Pydantic Settings validation (development/staging/production).
        These are intentionally separate to allow flexible naming in each system.
    """
    global _gsm_config
    if _gsm_config is None:
        project_id = os.getenv("GCP_PROJECT_ID", "my-hibachi-crm")
        # Use GSM_ENVIRONMENT for secret prefix, separate from ENVIRONMENT (Pydantic enum)
        # Default to "prod" to match existing GSM secrets (prod-backend-api-*)
        gsm_environment = os.getenv("GSM_ENVIRONMENT", "prod")
        _gsm_config = GSMConfig(project_id, gsm_environment)
    return _gsm_config


async def load_config() -> Dict[str, Any]:
    """
    Load complete configuration from GSM with environment fallbacks

    Returns:
        Dictionary with all configuration values
    """
    gsm = get_gsm_config()

    # Define all required secrets with their categories and fallback env vars
    required_secrets = [
        # Global secrets
        ("global", "CONFIG_VERSION", "CONFIG_VERSION"),
        ("global", "STRIPE_SECRET_KEY", "STRIPE_SECRET_KEY"),
        ("global", "OPENAI_API_KEY", "OPENAI_API_KEY"),
        ("global", "GOOGLE_MAPS_SERVER_KEY", "GOOGLE_MAPS_SERVER_KEY"),
        # Backend API secrets
        ("backend-api", "JWT_SECRET", "JWT_SECRET"),
        ("backend-api", "ENCRYPTION_KEY", "ENCRYPTION_KEY"),
        ("backend-api", "DB_URL", "DATABASE_URL"),
        ("backend-api", "REDIS_URL", "REDIS_URL"),
        ("backend-api", "RC_CLIENT_ID", "RINGCENTRAL_CLIENT_ID"),
        ("backend-api", "RC_CLIENT_SECRET", "RINGCENTRAL_CLIENT_SECRET"),
        ("backend-api", "DEEPGRAM_API_KEY", "DEEPGRAM_API_KEY"),
        ("backend-api", "SMTP_PASSWORD", "SMTP_PASSWORD"),
        ("backend-api", "GMAIL_APP_PASSWORD", "GMAIL_APP_PASSWORD"),
        # Cloudflare R2 Storage secrets
        ("backend-api", "cloudflare-r2-account-id", "CLOUDFLARE_R2_ACCOUNT_ID"),
        ("backend-api", "cloudflare-r2-access-key-id", "CLOUDFLARE_R2_ACCESS_KEY_ID"),
        ("backend-api", "cloudflare-r2-secret-access-key", "CLOUDFLARE_R2_SECRET_ACCESS_KEY"),
        ("backend-api", "cloudflare-r2-bucket", "CLOUDFLARE_R2_BUCKET"),
        ("backend-api", "cloudflare-r2-endpoint", "CLOUDFLARE_R2_ENDPOINT"),
        # Google Calendar integration (service account JSON)
        ("backend-api", "google-calendar-service-account", "GOOGLE_SERVICE_ACCOUNT_JSON"),
    ]

    logger.info("🔑 Loading configuration from GSM...")
    config = await gsm.get_secrets_batch(required_secrets)

    # Validate critical secrets
    critical_secrets = ["JWT_SECRET", "ENCRYPTION_KEY"]
    missing_critical = [key for key in critical_secrets if key not in config]

    if missing_critical:
        raise ValueError(f"Critical secrets missing: {missing_critical}")

    logger.info(f"✅ Loaded {len(config)} configuration values")
    return config


async def start_config_monitoring():
    """Start background monitoring for configuration changes"""
    logger.info("🔄 Configuration monitoring started")
    # This could be enhanced to watch for GSM changes
    pass
